src/AbstractSubscription.js

```diff
@@ -0,0 +1,125 @@
+import EncryptionUtil from './EncryptionUtil'
+
+const { OrderingUtil } = Utils
+const debug = debugFactory('StreamrClient::AbstractSubscription')
+
+let subId = 0
+function generateSubscriptionId() {
+    const id = subId
+    subId += 1
+    return id.toString()
+}
+
+const DEFAULT_PROPAGATION_TIMEOUT = 5000
+const DEFAULT_RESEND_TIMEOUT = 5000
+
+export default class AbstractSubscription extends EventEmitter {
+    constructor(
+        streamId, streamPartition, callback, groupKeys,
+        propagationTimeout = DEFAULT_PROPAGATION_TIMEOUT, resendTimeout = DEFAULT_RESEND_TIMEOUT, isCombined = false
+    ) {
+        super()
+        if (!streamId) {
+            throw new Error('No stream id given!')
+        }
+
+        if (!callback) {
+            throw new Error('No callback given!')
+        }
+        this.streamId = streamId
+        this.streamPartition = streamPartition
+        this.id = generateSubscriptionId()
+        this.groupKeys = groupKeys || {}
+
+        if (!isCombined) {
+            this.orderingUtil = new OrderingUtil(streamId, streamPartition, (orderedMessage) => {
+                const newGroupKey = EncryptionUtil.decryptStreamMessage(orderedMessage, this.groupKeys[orderedMessage.getPublisherId()])
+                if (newGroupKey) {
+                    this.groupKeys[orderedMessage.getPublisherId()] = newGroupKey
+                }
+                callback(orderedMessage.getParsedContent(), orderedMessage)
+                if (orderedMessage.isByeMessage()) {
+                    this.emit('done')
+                }
+            }, (from, to, publisherId, msgChainId) => {
+                this.emit('gap', from, to, publisherId, msgChainId)
+            }, propagationTimeout, resendTimeout)
+        }
+        this.state = AbstractSubscription.State.unsubscribed
+    }
+
+    _clearGaps() {
+        this.orderingUtil.clearGaps()
+    }
+
+    stop() {
+        this._clearGaps()
+    }
+
+    getState() {
+        return this.state
+    }
+
+    setState(state) {
+        debug(`Subscription: Stream ${this.streamId} state changed ${this.state} => ${state}`)
+        this.state = state
+        this.emit(state)
+    }
+
+    handleError(err) {
+        /**
+         * If parsing the (expected) message failed, we should still mark it as received. Otherwise the
+         * gap detection will think a message was lost, and re-request the failing message.
+         */
+        if (err instanceof Errors.InvalidJsonError && err.streamMessage) {
+            this.orderingUtil.markMessageExplicitly(err.streamMessage)
+        }
+        this.emit('error', err)
+    }
+
+    async _catchAndEmitErrors(fn) {
+        try {
+            return await fn()
+        } catch (err) {
+            console.error(err)
+            this.emit('error', err)
+            // Swallow rejection
+            return Promise.resolve()
+        }
+    }
+
+    static async validate(msg, verifyFn) {
```

Comment:
This seems a bit off here, perhaps. Could this logic be moved into the actual `verifyFn` which seems to be primarily `stream.verifyStreamMessage`?

Reply:
Why doesn't it belong here? This code was in `Subscription.js`, I just moved it in its separate function to avoid duplication since it is used in two different methods.
Moving this code to `verifyFn` would increase the size of `StreamrClient.js` which is already quite big.

Reply:
The last statements of the function seem fine to me. What I have trouble understanding are the first two checks.

```js
if (msg.version !== 31) {
     throw new Error(`Can handle only StreamMessageV31, not version ${msg.version}`)
}
```
There doesn't seem to be anything inherit about this function that makes it incompatible with other versions. Isn't it the `verifyFn` function itself the one who knows which versions it is or is not compatible with? What if we were passing this function an alternative `verifyFn` that _is_ compatible with other versions? Why would we want to prevent its usage here?

```js
if (msg.prevMsgRef == null) {
    debug('handleMessage: prevOffset is null, gap detection is impossible! message: %o', msg)
}
```
Isn't gap detection done mostly in `OrderingUtil`? If I saw this log message, I would expect to find the emitting debug statement somewhere in `OrderingUtil` or its related classes.